### PR TITLE
[IMP] sale : pay down payment

### PR DESCRIPTION
**Pull request by Feyensv**
Description of the issue/feature this PR addresses:  Partial payment amount to confirm quotations 

Current behavior before PR: Quotation auto-confirmation through e-sign and/or full payment

Desired behavior after PR is merged: Quotation auto-confirmation through e-sign and/or full/**down** payment 

* **sale**:
   * Percentage of payment required in Settings if Online confirmation through payment is activated (% of quotation total amount, default = 100%)
   * Application in mail templates and quotation portal 
* **payment** : Display of the amount of transactions (for manual acquirers)
* **sale_management** : Quotation templates support of required payment percentage

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale/models/res_company.py [hunk 1]
```diff
@@ -10,6 +10,7 @@ class ResCompany(models.Model):
     portal_confirmation_sign = fields.Boolean(string='Online Signature')
     portal_confirmation_pay = fields.Boolean(string='Online Payment')
     quotation_validity_days = fields.Integer(default=30, string="Default Quotation Validity (Days)")
+    quotation_confirmation_percentage = fields.Float(string="Requested pay down percentage for quotation confirmation", default=100.0)
```
Comment (jem-odoo) [inline addons/sale/models/res_company.py]: Maybe add an SQL constraint to force the percentage to be in [0,100]

File: addons/sale/models/sale.py [hunk 2]
```diff
@@ -36,6 +36,9 @@ def _get_default_require_signature(self):
     def _get_default_require_payment(self):
         return self.env.user.company_id.portal_confirmation_pay
 
+    def _get_default_required_payment_percentage(self):
```
Comment (jem-odoo) [inline addons/sale/models/sale.py]: guidelines: `_default_<field_name>`

File: addons/sale/models/sale.py [hunk 3]
```diff
@@ -238,8 +245,8 @@ def _compute_remaining_validity_days(self):
 
     @api.depends('transaction_ids')
     def _compute_authorized_transaction_ids(self):
-        for trans in self:
-            trans.authorized_transaction_ids = trans.transaction_ids.filtered(lambda t: t.state == 'authorized')
+        for so in self:
```
Comment (jem-odoo) [inline addons/sale/models/sale.py]: rather `order` or `sale_order`. 2 letters variables should be avoid

File: addons/sale/models/sale.py [hunk 4]
```diff
@@ -248,6 +255,11 @@ def _compute_amount_undiscounted(self):
             total += line.price_subtotal + line.price_unit * ((line.discount or 0.0) / 100.0) * line.product_uom_qty  # why is there a discount in a field named amount_undiscounted ??
         self.amount_undiscounted = total
 
+    @api.depends('required_payment_percentage', 'amount_total')
+    def _compute_confirmation_amount(self):
+        for record in self:
```
Comment (jem-odoo) [inline addons/sale/models/sale.py]: rather `order` or `sale_order`. `record` is more for mixin, when you don't know on which object you are working.

File: addons/sale/models/sale.py [hunk 5]
```diff
@@ -248,6 +255,11 @@ def _compute_amount_undiscounted(self):
...
+            self.confirmation_amount = self.amount_total*self.required_payment_percentage/100.0
```
Comment (jem-odoo) [inline addons/sale/models/sale.py]: no `self` in the loop.

File: addons/sale_management/models/sale_order_template.py [hunk 6]
```diff
@@ -16,6 +16,9 @@ def _get_default_require_signature(self):
     def _get_default_require_payment(self):
         return self.env.user.company_id.portal_confirmation_pay
 
+    def _get_default_required_payment_percentage(self):
```
Comment (jem-odoo) [inline addons/sale_management/models/sale_order_template.py]: guidelines: `_default_<field_name>`

File: addons/website_sale/security/website_sale.xml [hunk 7]
```diff
@@ -23,6 +23,12 @@
         <field name="perm_unlink" eval="False"/>
     </record>
 
+    <record id="payment_token_salesman_rule" model="ir.rule">
```
Comment (jem-odoo) [inline addons/website_sale/security/website_sale.xml]: guidelines: `<model_name>_rule_<group_name>`

File: addons/sale/models/sale.py (hunk 2 above)
Comment (Feyensv) [reply addons/sale/models/sale.py]: Do i also change the two other default methods (for require signature and require payment) while i'm on it?

Comment (jem-odoo) [reply addons/sale/models/sale.py]: No, no useless change. Moreover, it's not the purpose of the branch ;)

File: addons/sale/models/payment.py [hunk 8]
```diff
@@ -146,18 +146,20 @@ def action_view_sales_orders(self):
...
         if render_values:
             values.update(render_values)
+        if not amount:
+            amount = order.amount_total
         # Not very elegant to do that here but no choice regarding the design.
         self._log_payment_transaction_sent()
         return self.acquirer_id.with_context(submit_class='btn btn-primary', submit_txt=submit_txt or _('Pay Now')).sudo().render(
             self.reference,
-            order.amount_total,
+            amount,
```
Comment (dbeguin) [inline addons/sale/models/payment.py]: you can directly do a ternary expression : 
`amount if amount else order.amount_total,`

File: addons/sale/models/res_config_settings.py [hunk 9]
```diff
@@ -127,6 +130,11 @@ def _onchange_quotation_validity_days(self):
                 'warning': {'title': "Warning", 'message': "Quotation Validity is required and must be greater than 0."},
             }
 
+    @api.onchange('quotation_confirmation_percentage')
+    def _onchange_quotation_confirmation_percentage(self):
+        if self.quotation_confirmation_percentage <= 0 or self.quotation_confirmation_percentage > 100.0:
```
Comment (dbeguin) [inline addons/sale/models/res_config_settings.py]: with constraint on min/max quotation_confirmation_percentage, you won't need this onchange

File: addons/sale/models/res_company.py (hunk 1 above)
Comment (dbeguin) [reply addons/sale/models/res_company.py]: That way you won't need the sql constraint 'confirmation_percentage' in sale.py.

File: addons/sale/models/sale.py [hunk 10]
```diff
@@ -501,6 +529,11 @@ def _create_invoices(self, grouped=False, final=False):
         invoices_name = {}
 
         for order in self:
+            if order.has_to_be_down_paid():
```
Comment (tde-banana-odoo) [inline addons/sale/models/sale.py]: I am not sure it should be done here. Shouldn't the caller check that it is effectively in a case where invoices can be created before calling _create_invoices instead of simply skipping it without notifying anybody ? Purpose of this method is to create invoice based on SO. I think caller should ensure it is doable. This method should have standard I/O .

File: addons/sale/models/sale.py [hunk 11]
```diff
@@ -821,7 +857,7 @@ def _create_payment_transaction(self, vals):
             vals['acquirer_id'] = acquirer.id
 
         vals.update({
-            'amount': sum(self.mapped('amount_total')),
+            'amount': sum(self.mapped(lambda sol: sol.confirmation_amount if sol.has_to_be_down_paid() else sol.amount_total)),
```
Comment (tde-banana-odoo) [inline addons/sale/models/sale.py]: I think a list comprehension will be easier to understand here (and probably better for performances).

File: addons/sale/views/sale_portal_templates.xml [hunk 12]
```diff
@@ -331,7 +344,7 @@
                         </t>
                         <div t-elif="sale_order.has_to_be_paid(True)" class="col-sm-auto mt8">
                             <a role="button" class="btn btn-primary" data-toggle="modal" data-target="#modalaccept" href="#">
-                                <i class="fa fa-check"/> <t t-if="not sale_order.signature">Accept &amp; Pay</t><t t-else="">Pay now</t>
+                                <i class="fa fa-check"/> <t t-if="not sale_order.signature or sale_order.require_payment">Accept &amp; Pay</t><t t-else="">Pay now</t>
```
Comment (tde-banana-odoo) [inline addons/sale/views/sale_portal_templates.xml]: What is the purpose of this ? As we are already in an "has_to_be_paid" condition this should normally not be necessary. Am I wrong ?

File: addons/sale_management/models/sale_order_template.py [hunk 13]
```diff
@@ -24,6 +27,7 @@ def _get_default_require_payment(self):
         help='Number of days for the validity date computation of the quotation')
     require_signature = fields.Boolean('Online Signature', default=_get_default_require_signature, help='Request a online signature to the customer in order to confirm orders automatically.')
     require_payment = fields.Boolean('Online Payment', default=_get_default_require_payment, help='Request an online payment to the customer in order to confirm orders automatically.')
+    required_payment_percentage = fields.Float('Payment Amount', default=_default_required_payment_percentage, help='Percentage of payment requested to the customer in order to confirm orders automatically.')
```
Comment (tde-banana-odoo) [inline addons/sale_management/models/sale_order_template.py]: Note: just update fields accordingly of other fields (require_confirmation_percentage if this name is chosen for example).

File: addons/payment/views/payment_portal_templates.xml [hunk 14]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <odoo>
-    <data>
+
```
Comment (tde-banana-odoo) [inline addons/payment/views/payment_portal_templates.xml]: You may keep the data as this is an unnecessary diff to remove it.

File: addons/sale/models/sale.py (hunk 10 above)
Comment (tde-banana-odoo) [reply addons/sale/models/sale.py]: cc @jem-odoo the man who knows everything about invoice / so confirmation (or not, maybe)

File: addons/sale/models/sale.py [hunk 15]
```diff
@@ -142,6 +145,10 @@ def _get_payment_type(self):
     require_payment = fields.Boolean('Online Payment', default=_get_default_require_payment, readonly=True,
         states={'draft': [('readonly', False)], 'sent': [('readonly', False)]},
         help='Request an online payment to the customer in order to confirm orders automatically.')
+    required_payment_percentage = fields.Float('Payment Amount', default=_default_required_payment_percentage, readonly=True,
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/sale/models/sale.py]: @jem-odoo said the contrary in his review (it was _get_default previously) :-/

File: addons/sale/models/sale.py [hunk 16]
```diff
@@ -355,6 +373,16 @@ def _onchange_commitment_date(self):
                 }
             }
 
+    @api.onchange('require_payment')
+    def _onchange_require_payment(self):
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/sale/models/sale.py]: Maybe it isn't clear in the specs, but as I understood the task, the user should still be able to change manually on a SO whether it has to be confirmed with a payment (partial or not).

File: addons/sale/views/sale_portal_templates.xml (hunk 12 above)
Comment (Feyensv) [reply addons/sale/views/sale_portal_templates.xml]: It was to say that the payment was also sometimes used as only way of confirmation and that "Accept" should also be visible in this case, but the condition is wrong for that in its current state.

File: addons/sale/models/sale.py (hunk 10 above)
Comment (jem-odoo) [reply addons/sale/models/sale.py]: Yes, I agree with tde ! Otherwise, invoice for given SO will not be created and the user is not warned. I think we can say that all record in `self` should be invoicable, as prerequisite.

Comment (Feyensv) [reply addons/sale/models/sale.py]: Two cases : 
1) Draft/Sent : if the user pays for the SO, it is to confirm it and as the specs says, no invoice should be generated for the partial payment of confirmation
2) Confirmed : has_to_be_paid/down_paid == False and invoices are always generated
Invoices are never generated for unconfirmed quotations (at least not with the current UI).

Therefore, I only avoid the case where "automatic invoice" is checked and a payment is done for the SO so that it doesn't generate an invoice (a warning is added in the settings on this subject).

@jem-odoo Where could I catch the invoice creation otherwise when a partial confirmation payment is done (as fgi wants to avoid having invoices in this case).

Comment (Feyensv) [reply addons/sale/models/sale.py]: Moved to payment.py in `_invoice_sale_orders` to avoid modifying the general `_create_invoices` in sale.py.

File: addons/sale/models/sale.py [hunk 17]
```diff
@@ -36,6 +36,9 @@ def _get_default_require_signature(self):
     def _get_default_require_payment(self):
         return self.env.user.company_id.portal_confirmation_pay
 
+    def _default_required_payment_percentage(self):
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/sale/models/sale.py]: Just pushed a version with default value taken from company only when require_payment = True (in onchange and in default_get).
It potentially loosen the data on the SO but the code complexity is higher (or I didn't understand correctly the way you wanted me to do it).

**Comment by AntoineVDV** [comment]
Task cancelled
